File: client/lib/data/repositories/assignments/operations/create_assignment.dart
import 'package:dartz/dartz.dart';
import 'package:likha/core/errors/failures.dart';
import 'package:likha/core/sync/mutation_result.dart';
import 'package:likha/core/sync/sync_queue.dart';
import 'package:likha/core/utils/typedef.dart';
import 'package:likha/data/datasources/local/assignments/assignment_local_datasource.dart';
import 'package:likha/data/models/assignments/assignment_model.dart';
import 'package:likha/domain/assignments/entities/assignment.dart';
import 'package:uuid/uuid.dart';

ResultFuture<MutationResult<Assignment>> createAssignment(
  AssignmentLocalDataSource localDataSource,
  SyncQueue syncQueue,
  {
  required String classId,
  required String title,
  required String instructions,
  required int totalPoints,
  required bool allowsTextSubmission,
  required bool allowsFileSubmission,
  String? allowedFileTypes,
  int? maxFileSizeMb,
  required String dueAt,
  bool isPublished = true,
  int? termNumber,
  String? component,
  bool? noSubmissionRequired,
}) async {
  try {
    final assignmentId = const Uuid().v4();
    final queueEntryId = const Uuid().v4();
    final now = DateTime.now();

    final optimisticModel = AssignmentModel(
      id: assignmentId,
      classId: classId,
      title: title,
      instructions: instructions,
      totalPoints: totalPoints,
      allowsTextSubmission: allowsTextSubmission,
      allowsFileSubmission: allowsFileSubmission,
      allowedFileTypes: allowedFileTypes,
      maxFileSizeMb: maxFileSizeMb,
      dueAt: DateTime.parse(dueAt),
      isPublished: isPublished,
      orderIndex: 0,
      submissionCount: 0,
      gradedCount: 0,
      termNumber: termNumber,
      component: component,
      createdAt: now,
      updatedAt: now,
      syncStatus: SyncStatus.pending,
    );

    final data = {
      ...optimisticModel.toPayload(),
      if (noSubmissionRequired != null) 'no_submission_required': noSubmissionRequired,
    };

    final db = await localDataSource.localDatabase.database;
    await db.transaction((txn) async {
      await localDataSource.insertAssignment(optimisticModel, txn: txn);
      await syncQueue.enqueue(
        SyncQueueEntry(
          id: queueEntryId,
          entityType: SyncEntityType.assignment,
          operation: SyncOperation.create,
          payload: data,
          status: SyncStatus.pending,
          retryCount: 0,
          maxRetries: 5,
          createdAt: now,
        ),
        txn: txn,
      );
    });

    return Right(MutationResult(entity: optimisticModel, status: SyncStatus.pending));
  } catch (e) {
    return Left(ServerFailure(e.toString()));
  }
}

File: server/src/modules/assignment/repository_operations/create_assignment.rs
use chrono::Utc;
use sea_orm::*;
use uuid::Uuid;

use crate::utils::{AppError, AppResult};
use ::entity::assignments;

pub async fn create_assignment(
    db: &DatabaseConnection,
    class_id: Uuid,
    title: String,
    instructions: String,
    total_points: i32,
    allows_text_submission: bool,
    allows_file_submission: bool,
    allowed_file_types: Option<String>,
    max_file_size_mb: Option<i32>,
    due_at: chrono::NaiveDateTime,
    order_index: i32,
    client_id: Option<Uuid>,
    is_published: bool,
    term_number: Option<i32>,
    component: Option<String>,
) -> AppResult<assignments::Model> {
    let assignment = assignments::ActiveModel {
        id: Set(client_id.unwrap_or_else(Uuid::new_v4)),
        class_id: Set(class_id),
        title: Set(title),
        instructions: Set(instructions),
        total_points: Set(total_points),
        allows_text_submission: Set(allows_text_submission),
        allows_file_submission: Set(allows_file_submission),
        allowed_file_types: Set(allowed_file_types),
        max_file_size_mb: Set(max_file_size_mb),
        due_at: Set(due_at),
        is_published: Set(is_published),
        order_index: Set(order_index),
        created_at: Set(Utc::now().naive_utc()),
        updated_at: Set(Utc::now().naive_utc()),
        deleted_at: Set(None),
        term_number: Set(term_number),
        component: Set(component),
    };

    assignment
        .insert(db)
        .await
        .map_err(|e| AppError::InternalServerError(format!("Failed to create assignment: {}", e)))
}
